Issue 262 pb dm examine constraints

moodymudskipper opened the pull request:
closes #262

We add a dependency on {progress} package

We do some minor refactoring to loop only once in `check_pk_constraints`.

We add a progress bar for `check_pk_constraints` and `check_pk_constraints`.

It would be possible to have only one progress bar but we'd need deeper refactoring.

The following example prints as before because {progress} doesn't kick in for short operations, but if I insert artifical waiting times we get the following:

![image](https://user-images.githubusercontent.com/18351714/119567565-c1e58080-bdac-11eb-9689-3c4657dafc4c.png)

krlmlr commented:
/style

krlmlr commented:
/style

krlmlr commented:
/style

krlmlr commented on DESCRIPTION:
```diff
@@ -49,6 +49,7 @@ Imports:
     magrittr,
     memoise,
     methods,
+    progress,
```
I wonder if we should add a `progress = NA` argument to the functions that implement a progress bar. This would allow moving the package to "Suggests". Behavior:

- If `NA`, hide in non-interactive mode, show in interactive mode, give a message if package not installed
- If `FALSE`, hide
- If `TRUE`, fail if package is not installed

What do you think?

krlmlr reviewed (commented):
Thanks, looks good!

moodymudskipper commented on R/examine-constraints.R:
```diff
@@ -126,9 +133,17 @@ check_fk_constraints <- function(dm) {
+    format = "  checking fk constraints [:bar] :percent in :elapsed",
+    total = nrow(fks_tibble), clear = FALSE, width = 60)
+
   fks_tibble %>%
```
I don't understand what you refer to

moodymudskipper commented:
I've implemented a `new_ticker` function, it works as you described.
It has a `progress = NA` argument to accommodate your last comment.
I've added a `progress = NA` argument to all functions that call directly or indirectly `new_ticker()`
An exception is the deprecated functions, that I thought we wouldn't want to alter, so I didn't touch their arguments and they use `progress = FALSE` so nothing changes.
{progress} was moved from Imports to Suggests
the progress argument was documented in `?dm_examine_constraints`

krlmlr commented:
/style

krlmlr commented on R/examine-constraints.R:
```diff
@@ -100,13 +103,16 @@ check_pk_constraints <- function(dm) {
     ))
   }
   table_names <- pull(pks, table)
+  columns     <- pull(pks, pk_col)
```
```suggestion
  table_names <- pks$table
  columns     <- pks$pk_col
```

krlmlr commented on R/examine-constraints.R:
```diff
@@ -119,16 +125,19 @@ check_pk_constraints <- function(dm) {
     mutate(fks, t1 = cts, t2 = pts) %>%
     select(t1, t1_name = child_table, colname = child_fk_cols, t2, t2_name = parent_table, pk = parent_key_cols)
+
+  ticker <- new_ticker("checking pk constraints", nrow(fks_tibble))
```
Same above.

```suggestion
  ticker <- new_ticker("checking pk constraints", nrow(fks_tibble), progress = progress)
```

krlmlr commented on R/zzx-deprecated.R:
```diff
@@ -403,7 +403,7 @@ cdm_learn_from_db <- function(dest) {
 cdm_check_constraints <- function(dm) {
   deprecate_soft("0.1.0", "dm::cdm_check_constraints()", "dm::dm_examine_constraints()")
-  dm_examine_constraints_impl(dm = dm)
+  dm_examine_constraints_impl(dm = dm, progress = FALSE)
```
Do we still need this?

moodymudskipper replied:
Is it a question for me ? progress is NA by default, so by setting to FALSE I make sure old behavior is unchanged. Or do you mean to hard deprecate the function ?

krlmlr replied:
I'd expect that tests run in noninteractive mode, so this would not be necessary. But we can leave it as is.

krlmlr commented on R/new_ticker.R:
```diff
@@ -0,0 +1,33 @@
+  # message or fail depending on `progress`
+
+  if (is.na(progress)) {
+    progress <- interactive()
+    if (progress && !requireNamespace("progress")) {
+      message("The 'progress' package must be installed in order to display progress bars.")
```
This package uses rlang idioms where approriate.

```suggestion
    progress <- is_interactive()
    if (progress && !is_installed("progress")) {
      inform("The 'progress' package must be installed in order to display progress bars.")
```

krlmlr reviewed (commented):
Thanks, let's do one more iteration.

krlmlr commented on R/new_ticker.R:
```diff
@@ -0,0 +1,33 @@
+  if (!progress) return(identity)
+
+  if (!requireNamespace("progress")) {
+    stop("The 'progress' package must be installed in order to display progress bars.")
```
Ultimately, this should be `abort()`. I wonder if we should implement

```r
check_suggested <- function(package, use, message) {
  if (!use) {
    return(FALSE)
  } else if (is.na(use)) {
    if (is_installed(package)) {
  ...
}
```

so that this pattern can be applied consistently and we can move more packages to "Suggests" if needed?

moodymudskipper replied:
good idea, it's also more readable this way, I've implemented it

moodymudskipper replied:
Once we merge this we should use `check_suggested()` in `bdm_render_graph()` to replace :

```
  if (!rlang::is_installed("DiagrammeR")) {
    stop("DiagrammeR package needed for this function to work. Please install it.",
      call. = FALSE
    )
  }
```

krlmlr commented:
Thanks!

krlmlr commented on R/examine-constraints.R:
```diff
@@ -119,16 +125,19 @@ check_pk_constraints <- function(dm) {
     mutate(fks, t1 = cts, t2 = pts) %>%
     select(t1, t1_name = child_table, colname = child_fk_cols, t2, t2_name = parent_table, pk = parent_key_cols)
+
+  ticker <- new_ticker("checking pk constraints", nrow(fks_tibble), progress = progress)
```
This needs to be "fk constraints"? Should we spell out "primary key" and "foreign key"?
